gtsam/sfm/ShonanAveraging.cpp

```diff
@@ -471,57 +471,86 @@ Sparse ShonanAveraging<d>::computeA(const Values &values) const {
   return Lambda - Q_;
 }
 
+/* ************************************************************************* */
+template <size_t d>
+Sparse ShonanAveraging<d>::computeA(const Matrix &S) const {
+  auto Lambda = computeLambda(S);
+  return Lambda - Q_;
+}
+
+/* ************************************************************************* */
+// Perturb the initial initialVector by adding a spherically-uniformly
+// distributed random vector with 0.03*||initialVector||_2 magnitude to
+// initialVector
+// ref : Part III. C, Rosen, D. and Carlone, L., 2017, September. Computational
+// enhancements for certifiably correct SLAM. In Proceedings of the
+// International Conference on Intelligent Robots and Systems.
+Vector perturb(const Vector &initialVector) {
+  // generate a 0.03*||x_0||_2 as stated in David's paper
+  int n = initialVector.rows();
+  Vector disturb = Vector::Random(n);
+  disturb.normalize();
+
+  double magnitude = initialVector.norm();
+  Vector perturbedVector = initialVector + 0.03 * magnitude * disturb;
+  perturbedVector.normalize();
+  return perturbedVector;
+}
+
 /* ************************************************************************* */
 /// MINIMUM EIGENVALUE COMPUTATIONS
-// Alg.6 from paper Distributed Certifiably Correct Pose-Graph Optimization, 
-// it takes in the certificate matrix A as input, the maxIterations and the 
-// minEigenvalueNonnegativityTolerance is set to 1000 and 10e-4 ad default, 
+// Alg.6 from paper Distributed Certifiably Correct Pose-Graph Optimization,
+// it takes in the certificate matrix A as input, the maxIterations and the
+// minEigenvalueNonnegativityTolerance is set to 1000 and 10e-4 ad default,
 // there are two parts
 // in this algorithm:
-// (1) 
+// (1) compute the maximum eigenpair (\lamda_dom, \vect{v}_dom) of A by power
+// method. if \lamda_dom is less than zero, then return the eigenpair. (2)
+// compute the maximum eigenpair (\theta, \vect{v}) of C = \lamda_dom * I - A by
+// accelerated power method. Then return (\lamda_dom - \theta, \vect{v}).
 static bool PowerMinimumEigenValue(
     const Sparse &A, const Matrix &S, double *minEigenValue,
```

Comment:
OK this looks pretty good now, but I think it might be a *bit* cleaner to use ```double &``` instead of ```double*``` for ```minEigenValue``` and ```size_t& ``` instead of ```size_t*``` for ```numIterations```

(I like to avoid pointers whenever I can -- lots of creative ways to cause horrible segfaults with these :-P)

Reply:
I changed this part. I'm not 100% sure if there is further consideration to use `double*` than `double &` and `size_t` because this form follows previous code by @dellaert . Could you take a look?